envoy/api/v2/route/route.proto

```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
   // * The regex *\d{3}* matches the value *123*
   // * The regex *\d{3}* does not match the value *1234*
   // * The regex *\d{3}* does not match the value *123.456*
-  google.protobuf.BoolValue regex = 3;
+  google.protobuf.BoolValue regex = 3 [deprecated = true];
+
+  // [#not-implemented-hide:]
+  // Specifies how the header match will be performed to route the request.
+  // If header_match_specifier is absent, a request that has the name header will match, regardless
+  // of the header's value.
+  oneof header_match_specifier {
+    // [#not-implemented-hide:]
+    // If specified, header match will be performed based on the value of the header.
+    string value_match = 4;
+
+    // [#not-implemented-hide:]
+    // If specified, this regex string is a regular expression rule which implies the entire request
+    // header value must match the regex. The rule will not match if only a subsequence of the
+    // request header value matches the regex. The regex grammar used in the value field is defined
+    // `here <http://en.cppreference.com/w/cpp/regex/ecmascript>`_.
+    //
+    // Examples:
+    //
+    // * The regex *\d{3}* matches the value *123*
+    // * The regex *\d{3}* does not match the value *1234*
+    // * The regex *\d{3}* does not match the value *123.456*
+    string regex_match = 5;
+
+    // [#not-implemented-hide:]
+    // If specified, header match will be performed based on range.
+    // The rule will match if the request header value is within this range.
+    // This only applies to integer header values.
+    HeaderValueRange range_match = 6;
```

Comment:
The type name can be RangeMatch. The match can be applied to any value. It doesn't need to be from header.

Reply:
`Interval` might be another name. It doesn't need to be used in a match capacity..

Reply:
Interval often misleads to time. Range is more accurate in this context.

Reply:
OK, then simply `Range`?

Reply:
Google API Design Guide has a specific section on the range representation. We should always use [start, end) to represent a range. The reason is to avoid English debate like, [lower, upper], [lower, high], [first, last]. It is better to have a single universal representation instead of many locally optimized one.

Do we really care about int64 vs double? Once someone starts to deal with numbers bigger than a billion, it is better to use double, 1000000000000 is much worse to write than 1e12. I don't think there is any business reason to limit the header value to integers.

Reply:
I would strongly recommend defining an `IntegerRange` or `Int64Range` and using it for this feature. In the data path there is no reason to be performing double conversions in this case.

Reply:
I would suggest naming the field as "integer_range" instead of just range. I would also like to see the use case as @htuch said..

Reply:
@rshriram some explanation here https://github.com/envoyproxy/envoy/issues/2418
A very specific use case: Service Fabric supports a concept of [ranged partitioning scheme ](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-concepts-partitioning#ranged-partitioning-scheme) for stateful services. 
An incoming request will be routed to a given partition of the stateful service if the partition key is in the specified range for that partition.

Reply:
This is cool! Could you just stick this in the comments for this field? It will greatly help in the documentation when people look at this option, and get a sense for what the real world usage is.